File: amlip_cpp/src/cpp/node/StatusNode.cpp
// Copyright 2022 Proyectos y Sistemas de Mantenimiento SL (eProsima).
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

/**
 * @file StatusNode.cpp
 */

#include <cpp_utils/Log.hpp>
#include <cpp_utils/exception/InconsistencyException.hpp>

#include <dds/Participant.hpp>
#include <dds/network_utils/topic.hpp>
#include <amlip_cpp/node/StatusNode.hpp>

namespace eprosima {
namespace amlip {
namespace node {

StatusNode::StatusNode(
        const char* name,
        uint32_t domain_id)
    : ParentNode(name, types::NodeKind::status, types::StateKind::stopped, domain_id)
    , status_reader_(participant_->create_reader<types::StatusDataType>(
                dds::utils::STATUS_TOPIC_NAME,
                dds::utils::status_reader_qos()))
    , processing_(false)
{
    logInfo(AMLIPCPP_NODE_STATUS, "Created new Status Node: " << *this << ".");
}

StatusNode::StatusNode(
        const char* name)
    : StatusNode(name, dds::Participant::default_domain_id())
{
}

StatusNode::StatusNode(
        const std::string& name)
    : StatusNode(name.c_str())
{
}

StatusNode::~StatusNode()
{
    logDebug(AMLIPCPP_NODE_STATUS, "Destroying Status Node: " << *this << ".");

    // If the thread is running, stop it and join thread
    if (processing_)
    {
        stop_processing();
    }

    logDebug(AMLIPCPP_NODE_STATUS, "Status Node Destroyed.");
}

void StatusNode::process_status_async(
        const std::function<void(const types::StatusDataType&)>& callback)
{
    logInfo(AMLIPCPP_NODE_STATUS, "Start processing Status messages by callback in : " << *this << ".");

    if (processing_)
    {
        throw utils::InconsistencyException(
                  STR_ENTRY << "Status node " << this << " is already processing data.");
    }
    else
    {
        processing_ = true;
        process_thread_ = std::thread(&StatusNode::process_routine_, this, callback);

        change_status_(types::StateKind::running);
    }
}

void StatusNode::process_status_async(
        const StatusListener& callback_functor)
{
    logInfo(AMLIPCPP_NODE_STATUS, "Start processing Status messages by listener in : " << *this << ".");

    if (processing_)
    {
        throw utils::InconsistencyException(
                  STR_ENTRY << "Status node " << this << " is already processing data.");
    }
    else
    {
        processing_ = true;
        process_thread_ = std::thread(
            &StatusNode::process_routine_,
            this,
            [&callback_functor](const types::StatusDataType& status)
            {
                callback_functor.status_received(status);
            });

        change_status_(types::StateKind::running);
    }
}

void StatusNode::stop_processing()
{
    if (processing_)
    {
        processing_ = false;
        status_reader_->awake_waiting_threads(); // This must awake thread and it must finish
        process_thread_.join();

        change_status_(types::StateKind::stopped);
    }
}

void StatusNode::process_routine_(
        const std::function<void(const types::StatusDataType&)>& callback)
{
    while (processing_)
    {
        // Wait for data
        utils::event::AwakeReason reason = status_reader_->wait_data_available();

        if (reason == utils::event::AwakeReason::disabled)
        {
            logDebug(AMLIPCPP_NODE_STATUS, "Status Node " << *this << " finished processing data.");

            // Break thread execution
            return;
        }

        // Read data
        types::StatusDataType status = status_reader_->read();

        logDebug(AMLIPCPP_NODE_STATUS, "Status Node " << *this << " read data :" << status << ".");

        // Call callback
        callback(status);
    }
}

std::ostream& operator <<(
        std::ostream& os,
        const StatusNode& node)
{
    os << "STATUS_NODE{" << node.id() << "}";
    return os;
}

} /* namespace node */
} /* namespace amlip */
} /* namespace eprosima */

File: amlip_py/amlip_py/node/StatusNode.py
# Copyright 2022 Proyectos y Sistemas de Mantenimiento SL (eProsima).
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
"""AML-IP Status Node API specification."""


from amlip_py.types.AmlipIdDataType import AmlipIdDataType
from amlip_py.types.StatusDataType import StatusDataType

from amlip_swig import StatusListener as cpp_StatusListener
from amlip_swig import StatusNode as cpp_StatusNode


class StatusListener(cpp_StatusListener):
    """
    Status Listener class.

    This object must execute status_received method with each Status message that is received
    from  StatusNode.
    Must be inherited from user in order to reimplement status_received method.
    """

    def status_received(
            self,
            status: StatusDataType):
        """
        Raise exception.

        Abstract method.
        This method should be reimplemented by child class.
        """
        raise NotImplementedError('StatusListener.status_received must be specialized before use.')


class StatusLambda(StatusListener):
    """
    Custom StatusListener supporting to create it with a lambda function.

    This object is created with a lambda function that is stored inside and used for every
    Status message received.
    """

    def __init__(self, callback):
        """Construct new object with lambda as callback."""
        self.callback_ = callback
        super().__init__()

    def status_received(
            self,
            status: StatusDataType):
        """Call internal lambda."""
        self.callback_(status)


class StatusNode(cpp_StatusNode):
    """
    AML-IP Status Node.

    This Node implements a Reader to read every Status message published in the AML network.
    With each message it executes the listener method (or lambda) given to process_status_async.

    This Node has an internal thread that listen to messages and execute the callback.
    In order to start processing messages, use process_status_async.
    In order to stop processing messages, use stop_processing.
    """

    def __init__(
            self,
            name: str):
        """
        Create a new Status Node with a given name.

        Parameters
        ----------
        name : str
            Name of the node.
        """
        #####
        # Internal variables
        self.internal_listener_ = None

        #####
        # Parent class constructor
        super().__init__(name)

    def process_status_async(
            self,
            callback=None,
            listener: StatusListener = None) -> None:
        """
        Start processing the Status data arrived by using a lambda.

        This method supports 2 ways of including the lambda that describes what to do
        with each Status data that arrives. One is by a listener that may specialize method
        process_job and the other is retrieving a lambda as callback.

        Preconditions
        -------------
            One and only one of this arguments must be set

        Parameters
        ----------
        callback : lambda (StatusDataType)
            Lambda which only argument is a Status
            [Default = None]

        listener : StatusListener
            Object that specializes status_received method to execute in message reception.
            [Default = None]

        """
        # If both arguments are set or neither of them, fail
        if (listener and callback):
            raise ValueError(
                'Method process_status_async expects a listener object or a callback, both given.')

        # In case one and only one argument given, store it internally as a StatusListener
        # so this object is not destroyed while used.
        elif listener:
            self.internal_listener_ = listener
            return cpp_StatusNode.process_status_async(self, self.internal_listener_)

        elif callback:
            self.internal_listener_ = StatusLambda(callback)
            return cpp_StatusNode.process_status_async(self, self.internal_listener_)

        else:
            raise ValueError(
                'Method process_status_async expects a listener object or a callback, none given.')

    def stop_processing(
            self) -> None:
        """Stop this entity if it is processing data. Do nothing otherwise."""
        cpp_StatusNode.stop(self)

        # Unstore the internal Listener
        self.internal_listener_ = None

    def get_id(
            self) -> AmlipIdDataType:
        """Get AMLIP id of the node."""
        return cpp_StatusNode.id(self)
